### PR TITLE
Feature/465 directed graphsage

### PR DESCRIPTION
**Pull request by geoffj-d61**
Updated the GraphSAGE aggregation architecture to create a directed GraphSAGE algorithm that separately handles in-node and out-node neighbourhoods. Some (temporary) demo examples have been added to check the differences in behaviour between directed and undirected GraphSAGE. 

Note that example 3 (directed GraphSAGE, no in-nodes, only out-nodes) is now broken; this is not because of directionality, but because of an existing bug in the framework: the mean of a tensor with a 0 dimension produces NaNs. Undirected GraphSAGE suffers this same bug (try setting the 2nd sampling dimension to 0 in either example 1 or 2). Will need help to fix this.

### CODE REVIEW

**Comment by adocherty** [comment]
OK, my changes have broken the unit tests. This is due to the `AttentionalAggregator` which doesn't fit into the paradigm of the other aggregators. I'll commit a fix tomorrow 😟

**Comment by geoffj-d61** [comment]
@adocherty I have moved example4 into graphsage, as you suggested, and deleted directed-graphsage. The example and experimental notebooks have been saved as branch feature/465-directed-graphsage on stellar-research.

File: stellargraph/layer/graphsage.py [hunk 1]
```diff
@@ -127,8 +135,39 @@ def build(self, input_shape):
                 trainable=True,
             )
 
+        # If there are zero neighbours, back off to
+        # using only node features.
+        counter = 1
+        for neigh_shape in input_shape[1:]:
+            if neigh_shape[2] == 0:
+                counter = counter + 1
+        self._build_mlp_only = counter == len(input_shape)
```
[the remark this answers is not in the document]
Comment (geoffj-d61) [reply stellargraph/layer/graphsage.py]: Method _build_mlp_only() removed.

File: stellargraph/layer/graphsage.py [hunk 2]
```diff
@@ -302,7 +352,7 @@ def build(self, input_shape):
                 trainable=True,
             )
 
-    def aggregate_neighbours(self, x_neigh):
+    def aggregate_neighbours(self, x_neigh, **kwargs):
```
[the remark this answers is not in the document]
Comment (geoffj-d61) [reply stellargraph/layer/graphsage.py]: Fixed by @adocherty .

File: stellargraph/layer/graphsage.py [hunk 3]
```diff
@@ -747,3 +806,198 @@ def default_model(self, flatten_output=True):
...
+            or self.out_samples is None
+            or self.input_feature_size is None
+        ):
+            raise ValueError(
+                "If generator is not provided, in_samples, out_samples and input_dim must be specified."
+            )
+
+    def _build_aggregators(self):
+        # Model parameters
+        self.max_hops = max_hops = self.n_layers
```
[the remark this answers is not in the document]
Comment (geoffj-d61) [reply stellargraph/layer/graphsage.py]: Renamed n_layers to max_slots.

File: stellargraph/layer/graphsage.py [hunk 4]
```diff
@@ -747,3 +806,198 @@ def default_model(self, flatten_output=True):
...
+        stage_tree = xin
+        for stage in range(self.max_hops):
+            stage_tree = aggregate_neighbours(stage_tree, stage)
+        out_layer = stage_tree[0]
+
+        # Remove neighbourhood dimension from output tensors of the stack
+        out_layer = Reshape(K.int_shape(out_layer)[2:])(out_layer)
+        return self._normalization(out_layer)
+
+    def _compute_input_sizes(self) -> List[int]:
```
[the remark this answers is not in the document]
Comment (geoffj-d61) [reply stellargraph/layer/graphsage.py]: Added call from `GraphSAGE.__init__()` to `_compute_neighbourhood_sizes()` to set `self.neighbourhood_sizes` (and `max_slots` for `DirectedGraphSAGE`). Updated `GraphSAGE.node_model()` to use `self.neighbourhood_sizes`, and removed `DirectedGraphSAGE.node_model()`.

File: stellargraph/layer/graphsage.py [hunk 5]
```diff
@@ -747,3 +806,198 @@ def default_model(self, flatten_output=True):
...
+        if (
+            self.in_samples is None
+            or self.out_samples is None
+            or self.input_feature_size is None
+        ):
+            raise ValueError(
+                "If generator is not provided, in_samples, out_samples and input_dim must be specified."
+            )
+
+    def _build_aggregators(self):
```
[the remark this answers is not in the document]
Comment (geoffj-d61) [reply stellargraph/layer/graphsage.py]: - Shifted setting of `max_slots` to `DirectedGraphSAGE._compute_neighbourhood_sizes()`. 

- Shifted checking of `in_samples` and `out_samples` lengths to `_get_sizes_from_generator()` and `_get_sizes_from_keywords()`. 

- Added similar checking to `GraphSAGE` versions. 

- Removed redundant `DirectedGraphSAGE._build_aggregators()` method.

File: stellargraph/data/explorer.py [hunk 6]
```diff
@@ -1008,106 +708,144 @@ def run(self, nodes=None, n=1, n_size=None, seed=None):
 
         return walks
 
-    def _check_parameter_values(self, nodes, n, n_size, graph_schema, seed):
+
+class DirectedBreadthFirstNeighbours(GraphWalk):
```
[the remark this answers is not in the document]
Comment (geoffj-d61) [reply stellargraph/data/explorer.py]: I guess this refactoring will be another issue.

File: stellargraph/layer/graphsage.py [hunk 7]
```diff
@@ -747,3 +806,198 @@ def default_model(self, flatten_output=True):
...
+            )
+
+        # Combine GraphSAGE layers in stages
+        stage_tree = xin
+        for stage in range(self.max_hops):
+            stage_tree = aggregate_neighbours(stage_tree, stage)
+        out_layer = stage_tree[0]
+
+        # Remove neighbourhood dimension from output tensors of the stack
+        out_layer = Reshape(K.int_shape(out_layer)[2:])(out_layer)
```
[the remark this answers is not in the document]
Comment (geoffj-d61) [reply stellargraph/layer/graphsage.py]: @adocherty what is `x` here in `if K.int_shape(x)==1`? The method input is `xin` which is the (flattened binary tree) list of tensors, so it's not that. And the `out_layer` we are expecting to have 2 or more dimensions. So I'm a bit confused here!

[the remark this answers is not in the document]
Comment (geoffj-d61) [reply stellargraph/layer/graphsage.py]: @adocherty I think the above is the last outstanding issue.

[the remark this answers is not in the document]
Comment (adocherty) [reply stellargraph/layer/graphsage.py]: Hi, yes I meant `K.int_shape(out_layer)[1] == 1`

[the remark this answers is not in the document]
Comment (geoffj-d61) [reply stellargraph/layer/graphsage.py]: Added and committed!

**Comment by geoffj-d61** [comment]
The one thing I forgot to check fully was the demo, which got broken by the keras -> tensorflow.keras changes! Fixed now.

**Pull request merged**